realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/SyncedRealmTests.java

```diff
@@ -0,0 +1,153 @@
+                .build();
+
+        Realm realm = null;
+        try {
+            realm = Realm.getInstance(config);
+            fail();
+        } catch (IllegalStateException ignored) {
+        } finally {
+            if (realm != null) {
+                realm.close();
+            }
+        }
+    }
+
+    // Login user on a worker thread, so this method can be used from both UI and non-ui threads.
+    @NonNull
+    private SyncUser loginUser() {
+        final CountDownLatch userReady = new CountDownLatch(1);
+        final AtomicReference<SyncUser> user = new AtomicReference<>();
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                SyncCredentials credentials = SyncCredentials.usernamePassword(UUID.randomUUID().toString(), "password", true);
+                user.set(SyncUser.login(credentials, Constants.AUTH_URL));
+                userReady.countDown();
+            }
+        }).start();
+        TestHelper.awaitOrFail(userReady);
+        return user.get();
+    }
+
+    @Test
+    public void waitForInitialRemoteData() {
+        // TODO We can improve this test once we got Sync Progress Notifications. Right now we cannot detect
+        // when a Realm has been uploaded.
+        SyncCredentials credentials = SyncCredentials.usernamePassword(UUID.randomUUID().toString(), "password", true);
+        SyncUser user = SyncUser.login(credentials, Constants.AUTH_URL);
+        SyncConfiguration config = new SyncConfiguration.Builder(user, Constants.USER_REALM)
+                .waitForInitialRemoteData()
+                .build();
+
+        Realm realm = null;
+        try {
+            realm = Realm.getInstance(config);
+            assertTrue(realm.isEmpty());
+        } finally {
+            if (realm != null) {
+                realm.close();
+            }
+        }
+    }
+
+    // This tests will start and cancel getting a Realm 10 times. The Realm should be resilient towards that
+    // We cannot do much better since we cannot control the order of events internally in Realm which would be
+    // needed to correctly test all error paths.
+    @Test
+    public void waitForInitialData_resilientInCaseOfRetries() throws InterruptedException {
+        SyncCredentials credentials = SyncCredentials.usernamePassword(UUID.randomUUID().toString(), "password", true);
+        SyncUser user = SyncUser.login(credentials, Constants.AUTH_URL);
+        final SyncConfiguration config = new SyncConfiguration.Builder(user, Constants.USER_REALM)
+                .waitForInitialRemoteData()
+                .build();
+        Random randomizer = new Random();
+
+        for (int i = 0; i < 10; i++) {
+            Thread t = new Thread(new Runnable() {
+                @Override
+                public void run() {
+                    Realm realm = null;
+                    try {
+                        realm = Realm.getInstance(config);
```

Comment:
i think it is actually possible that just call `Thread.currentThread().interrupt()` here. As long as the downloading latch is the first latch the `getInstance()` reaches, it will just throw there. So no need to loop 10 times.